File: PythonSpark/Aggregations/Purchases/PurchasesAggregationJob.py
from abc import ABC

from pyspark.sql import DataFrame

from Aggregations.AggregationJob import AggregationJob
from Utils.Loading import load_dynamic_table, load_static_table
from Utils.Parsing import filter_data_frame
from Utils.RichArgumentParser import RichArgumentParser


class PurchasesAggregationJob(AggregationJob, ABC):
    def load(self) -> dict:
        return {
            "purchases": load_dynamic_table("purchases", self.arguments, self.spark),
            "ips": load_static_table("ips", self.arguments, self.spark)
        }

    def process(self, data: dict) -> DataFrame:
        joined = data["purchases"].join(data["ips"], "ip")
        return filter_data_frame(joined,
                                 self.arguments.locations,
                                 self.arguments.time_frame,
                                 self.arguments.devices)

    @staticmethod
    def parser() -> RichArgumentParser:
        return AggregationJob.parser() \
            .add_dynamic_table_data_source("purchases") \
            .add_static_table_data_source("ips")

File: ScalaSpark/src/scala/Aggregations/Purchases/PurchasesAggregationJob.scala
package Aggregations.Purchases

import Aggregations.AggregationJob
import JobManagment.JobCompanion
import Utils.ArgumentsParsing.RichParser
import Utils.Loading.{loadDynamic, loadStatic}
import Utils.Parsing.filterData
import de.halcony.argparse.Parser
import org.apache.spark.sql.{DataFrame, SparkSession}

abstract class PurchasesAggregationJob(arguments: Map[String, String], spark: SparkSession) extends AggregationJob(arguments, spark) {
  override def load(): Map[String, DataFrame] = {
    Map(
      "purchases" -> loadDynamic("purchases", arguments, spark),
      "ips" -> loadStatic("ips", arguments, spark)
    )
  }

  override def process(data: Map[String, DataFrame]): DataFrame = {
    val purchasesWithLocations = data("purchases").join(data("ips"), "ip")

    filterData(arguments("timeFrame"), arguments("devices"), arguments("locations"), purchasesWithLocations)
  }
}

object PurchasesAggregationJob extends JobCompanion {
  override def parser(): Parser = {
    AggregationJob.parser()
      .addDynamicTableDataSource("purchases")
      .addStaticTableDataSource("ips")
  }
}
